// react-native-git-check/index.js
/**
 * Created by git on 17/4/7.
 * @flow
 */

'use strict';

import React, {PropTypes} from 'react';
import {View, StyleSheet, Text, TouchableOpacity, Image} from 'react-native';

class Item extends React.Component {
    static PropsType = {
        dataSource:PropTypes.string,
        onPress:PropTypes.func,
        multiSelect:PropTypes.bool
    }
    constructor(props, context)
    {
        super(props, context);
        this.state={
            selected:false,
        }
        this._onPress = this._onPress.bind(this)
    }

    componentWillReceiveProps(nextProps) {
        if(nextProps.onClickAll){   //点击了全选,根据全选的情况改变
            this.setState({
                selected: nextProps.clickAll
            });
        }else{   //点击了单个cell
            if(this.props.multiSelect === false){  //单选的情况
                this.setState({
                    selected:nextProps.dataSource === nextProps.radioValue
                })
            }
        }
    }
    render()
    {
        let {dataSource, selectedIcon, unselectedIcon, itemStyle}  = this.props
        let pic = this.state.selected?selectedIcon || require('./img/selected.png'):unselectedIcon || require('./img/unselected.png');
        let radio = <Image style={{marginLeft:10,alignSelf:'center'}} source={pic}/>
        return (
            <TouchableOpacity
                style={[{paddingVertical:10,flexDirection:'row',justifyContent:'space-between'},itemStyle]}
                onPress={this._onPress}>
                <Text>{dataSource.name}</Text>
                {radio}
            </TouchableOpacity>
        );
    }
    _onPress(){
        this.setState({selected:!this.state.selected})
        this.props.onPress && this.props.onPress(this.props.dataSource,!this.state.selected)
    }
}


class Check extends React.Component
{
    static propTypes={
        dataSource:React.PropTypes.array.isRequired,
        selectedIcon:Image.propTypes.source,
        unselectedIcon:Image.propTypes.source,
        style:View.propTypes.style,
        itemStyle:View.propTypes.style,
        topView:View.propTypes.style
    }
    constructor(props, context) {
        super(props, context);
        this.state={
            clickAll:false,
            onClickAll:false,
            dataSource:this.props.dataSource
        }
        this._onPress = this._onPress.bind(this)
    }
    _dataSource = []
    render(){
        let {dataSource, selectedIcon, unselectedIcon, itemStyle, topView}  = this.props
        let BottomView = dataSource.map((item,index)=>{
            return (

                <Item       dataSource={item}
                            //selected={dataSource.selected}
                            onPress={this._onPress}
                            multiSelect={true}
                            clickAll={this.state.clickAll}
                            onClickAll={this.state.onClickAll}
                            selectedIcon={selectedIcon}
                            unselectedIcon={unselectedIcon}
                            itemStyle={itemStyle}
                             key={index + 'check'}
                            />
            )
        })
        let pic = this.state.clickAll?selectedIcon || require('./img/selected.png'):unselectedIcon || require('./img/unselected.png');
        let radio = <Image style={{marginLeft:10,alignSelf:'center'}} source={pic}/>
        return (
            <View style={[styles.container,this.props.style]} >
                <View style={[styles.topView,topView]}>
                    <TouchableOpacity style={{flexDirection:'row'}} onPress={this._clickAll}>
                        <Text>全选</Text>
                        {radio}
                    </TouchableOpacity>
                </View>
                {BottomView}
            </View>
        );
    }
    _onPress(data,flag){
        this.setState({onClickAll:false})
        if(flag){
            this._dataSource.push(data)
            if(this._dataSource.length === this.state.dataSource.length){
                this.setState({clickAll:true})
            }
        }else{
            let index = this._dataSource.indexOf(data)
            this._dataSource.splice(index,1)
            this.setState({clickAll:false})
        }
        this.props.currentSelect && this.props.currentSelect(this._dataSource)
        //console.log( this._dataSource)
    }
    _clickAll = ()=>{
        this.setState({
            clickAll:!this.state.clickAll,
            onClickAll:true
        })
        this.forceUpdate()
        if(!this.state.clickAll){
            this.state.dataSource.map((data)=>{
                if(this._dataSource.indexOf(data) === -1){
                    this._dataSource.push(data)
                }
            })
        }else{
            this._dataSource = []
        }
        this.props.currentSelect && this.props.currentSelect(this._dataSource)
    }
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
    },
    topView:{paddingRight:5,marginLeft:5,flexDirection:'row',justifyContent:'flex-end',borderBottomWidth:StyleSheet.hairlineWidth,borderBottomColor:'#ddd',paddingVertical:5}
});

export default Check;
